### ISSUE TITLE
Some Stored Procedures not being generated in the Context file

### ISSUE THREAD

**Issue by VamsiKande** [ReversePOCO/EntityFramework-Reverse-POCO-Code-First-Generator#591]
Hi, 
I have few stored procedures under "xxxx." schema name, when using the schema filter or by include stored procedure Regex filter, a couple of stored procedures are being skipped under that schema, but rest of the stored procs are being generated with no issues under the same schema.

Tried filtering by both SchemaFilter and StoredProcedureFilter, both ways the specific strored procs are being skipped. Please help me out on resolving this issue.

**Comment by sjh37** [comment]
Hi. Can you list your lines of code you are using for filtering.

**Comment by VamsiKande** [comment]
Here are the settings

```c#
FilterSettings.IncludeStoredProcedures      = true;
FilterSettings.IncludeTableValuedFunctions  = true;
FilterSettings.IncludeScalarValuedFunctions = true;

FilterSettings.SchemaFilters.Add(new RegexIncludeFilter("Report.*"));
```
------------------------------------------------------------------------------------
I also tried by adding my specific stored proc that is being skipped 
```c#
FilterSettings.StoredProcedureFilters.Add(new RegexIncludeFilter("^uspInsert$"));
```

**Comment by sjh37** [comment]
Is the stored proc `uspInsert` using the `Report` schema?
I suspect not, hence why it's getting stripped out.
```c#
FilterSettings.SchemaFilters.Add(new RegexIncludeFilter("[Rr]eport.*"));
FilterSettings.SchemaFilters.Add(new RegexIncludeFilter("YourSchemaFor uspInsert"));
```

**Comment by VamsiKande** [comment]
Yes, the stored proc is using "Report" Schema.

I even tried with following Regex you provided, didn't work
FilterSettings.SchemaFilters.Add(new RegexIncludeFilter("[Rr]eport.*"));

**Comment by sjh37** [comment]
ok, that is odd. Let me investigate.
I have just tested this and restricted it to a single namespace, and I did get all the stored procedures and tables generated for that single namespace only, so is working ok.

Could you show the creation sql for the stored procedure, without the details in the middle.
for example:

```sql
CREATE PROCEDURE Alpha.Overclock (@Parameter DATETIME)
AS
BEGIN
    -- blah
END
```

**Comment by VamsiKande** [comment]
Sorry for getting back to you late, here is SP,

USE [Main]
GO

SET ANSI_NULLS ON
GO

SET QUOTED_IDENTIFIER ON
GO

create PROCEDURE [Report].[xxxxxx]
	@StartDate date,
	@EndDate date
AS
BEGIN
    --------------------
END
GO

**Comment by sjh37** [comment]
Thanks. Nothing strange in your SQL so all is ok.
It should be generating the code to call the SP's without being filtered out. Let me double check if there could be another reason things could be filtered out.

Could you also supply the name of a couple of stored procedures that are being filtered out. I wonder if there are any illegal characters in there.

**Comment by VamsiKande** [comment]
The following SP's are being filtered out, 

uspFreightInnovationsVolumeByDay
uspDailyDealsVolumeByDay
uspDailyDealsVolumeByClient

**Comment by sjh37** [comment]
Thanks. Again, nothing special with those names either.

**Comment by sjh37** [comment]
Try running this sql, see if those appear in the results.

```sql
SELECT  R.SPECIFIC_SCHEMA,
        R.SPECIFIC_NAME,
        R.ROUTINE_TYPE,
        R.DATA_TYPE as RETURN_DATA_TYPE,
        P.ORDINAL_POSITION,
        P.PARAMETER_MODE,
        P.PARAMETER_NAME,
        P.DATA_TYPE,
        ISNULL(P.CHARACTER_MAXIMUM_LENGTH, 0) AS CHARACTER_MAXIMUM_LENGTH,
        ISNULL(P.NUMERIC_PRECISION, 0) AS NUMERIC_PRECISION,
        ISNULL(P.NUMERIC_SCALE, 0) AS NUMERIC_SCALE,
        ISNULL(P.DATETIME_PRECISION, 0) AS DATETIME_PRECISION,
        P.USER_DEFINED_TYPE_SCHEMA + '.' + P.USER_DEFINED_TYPE_NAME AS USER_DEFINED_TYPE
FROM    INFORMATION_SCHEMA.ROUTINES R
        LEFT OUTER JOIN INFORMATION_SCHEMA.PARAMETERS P
            ON P.SPECIFIC_SCHEMA = R.SPECIFIC_SCHEMA
               AND P.SPECIFIC_NAME = R.SPECIFIC_NAME
WHERE   R.ROUTINE_TYPE = 'PROCEDURE'
        AND (
             P.IS_RESULT = 'NO'
             OR P.IS_RESULT IS NULL
            )
        AND R.SPECIFIC_SCHEMA + R.SPECIFIC_NAME IN (
            SELECT  SCHEMA_NAME(sp.schema_id) + sp.name
            FROM    sys.all_objects AS sp
                    LEFT OUTER JOIN sys.all_sql_modules AS sm
                        ON sm.object_id = sp.object_id
            WHERE   sp.type = 'P'
                    AND (CAST(CASE WHEN sp.is_ms_shipped = 1 THEN 1
                                   WHEN (
                                         SELECT major_id
                                         FROM   sys.extended_properties
                                         WHERE  major_id = sp.object_id
                                                AND minor_id = 0
                                                AND class = 1
                                                AND name = N'microsoft_database_tools_support'
                                        ) IS NOT NULL THEN 1
                                   ELSE 0
                              END AS BIT) = 0))

UNION ALL
SELECT  R.SPECIFIC_SCHEMA,
        R.SPECIFIC_NAME,
        R.ROUTINE_TYPE,
        R.DATA_TYPE as RETURN_DATA_TYPE,
        P.ORDINAL_POSITION,
        P.PARAMETER_MODE,
        P.PARAMETER_NAME,
        P.DATA_TYPE,
        ISNULL(P.CHARACTER_MAXIMUM_LENGTH, 0) AS CHARACTER_MAXIMUM_LENGTH,
        ISNULL(P.NUMERIC_PRECISION, 0) AS NUMERIC_PRECISION,
        ISNULL(P.NUMERIC_SCALE, 0) AS NUMERIC_SCALE,
        ISNULL(P.DATETIME_PRECISION, 0) AS DATETIME_PRECISION,
        P.USER_DEFINED_TYPE_SCHEMA + '.' + P.USER_DEFINED_TYPE_NAME AS USER_DEFINED_TYPE
FROM    INFORMATION_SCHEMA.ROUTINES R
        LEFT OUTER JOIN INFORMATION_SCHEMA.PARAMETERS P
            ON P.SPECIFIC_SCHEMA = R.SPECIFIC_SCHEMA
               AND P.SPECIFIC_NAME = R.SPECIFIC_NAME
WHERE   R.ROUTINE_TYPE = 'FUNCTION'
ORDER BY R.SPECIFIC_SCHEMA,
        R.SPECIFIC_NAME,
        R.ROUTINE_TYPE
```

**Comment by VamsiKande** [comment]
yes, those appeared in the results by running the above sql.

**Comment by sjh37** [comment]
So not sql server security related. Therefore it is **definitely** being filtered out in the generator.
I am at a loss, unless I can debug it myself. Any chance you can send me your files via a dropbox/onedrive link. or even a zip via email. simon@reversepoco.co.uk.
What I would like is your 

- `<database>.tt` settings file
- the SQL to create the stored procedures for the report schema.

**Comment by VamsiKande** [comment]
I sent an email of one of the stored proc that's filtered out and tt file to simon@reversepoco.co.uk.

**Comment by sjh37** [comment]
Received. Thanks, I'll check it out tomorrow.

**Comment by MarkLFT** [comment]
Not sure if related, but I also have a strange issue with including stored procedures.  If I add a single RegExIncludeFilter on the StoredProcedureFilter, it adds correctly.  If I then add a second filter, the second is not generated, but neither is the first.  Again, if I have just the second filter, it generates correctly.

These two strored procedures have no tables etc. in common.  And they are just two of about 50 stored procedures in the database.

The commands I am using are:

FilterSettings.StoredProcedureFilters.Add(new RegexIncludeFilter("spGetNextSequence.*")); 
FilterSettings.StoredProcedureFilters.Add(new RegexIncludeFilter("spGetOpenInvoices.*"));

If either row is commented out, the other works, but un-comment both, and neither works.

In addition I have found, if I donot use any filters, all SPs are generated correctly.  So it seems to be a filtering issue.

Any ideas of the cause of this?  Although there are around 50 SPs in this database, I only need to use about 5 in the data model, so I don't want to include them all.

**Comment by sjh37** [comment]
@MarkLFT I know exactly what the problem is. For now use a single include regex which include both your stored procedures. I'll add a fix for this.

**Comment by VamsiKande** [comment]
Hi, by any chance do we have any update on this issue?

**Comment by sjh37** [comment]
Hi @VamsiKande Should be fixed now.
Grab the latest `EF.Reverse.POCO.v3.ttinclude` file from [GitHub](https://raw.githubusercontent.com/sjh37/EntityFramework-Reverse-POCO-Code-First-Generator/master/EntityFramework.Reverse.POCO.Generator/EF.Reverse.POCO.v3.ttinclude) and replace your project file with this one.
Let me know if this fixes your problem.

**Comment by VamsiKande** [comment]
sure. Thanks!

**Comment by sjh37** [comment]
If this works, I'll make a new release tomorrow as this is a significant fix.
I'll pick this thread up in the morning.

**Comment by VamsiKande** [comment]
![Errors_New_ttinclude](https://user-images.githubusercontent.com/62710432/79396932-8ab66600-7f42-11ea-9126-6a53934ce2a0.PNG)

tried with the new ttinclude, seeing the errors as attached in the screen shot from the ttinclide file.

**Comment by sjh37** [comment]
Hi @VamsiKande I have double checked everything, and it runs ok. Can you re-check the ttinclude file was copied in ok.

**Comment by VamsiKande** [comment]
ok, I tested with the new tt file, the stored procedures are still excluded.

**Comment by VamsiKande** [comment]
Hi, do we have an update on this?

**Comment by sjh37** [comment]
Not yet, It tests ok for me. I will try again with a different database and try playing with the include filters some more. For now, place all your include filters in a single regex.

**Comment by VamsiKande** [comment]
Hi @sjh37 , do we have any update on this, do we have a solution to resolve the issue with certain SP are not being generated, you mentioned you resolved the issue but looks like it didn't, can you please let me know what's the difference with that particular stored procs that are not being generated?

**Comment by sjh37** [comment]
Re-checking again now...

**Comment by sjh37** [comment]
Still tests ok for me. I've added unit tests to test it works ok. Made a few tweaks. Grab the latest [EF.Reverse.POCO.v3.ttinclude](https://github.com/sjh37/EntityFramework-Reverse-POCO-Code-First-Generator/blob/master/EntityFramework.Reverse.POCO.Generator/EF.Reverse.POCO.v3.ttinclude) and copy over your `EF.Reverse.POCO.v3.ttinclude` in your project.

Could you also email me a DDL of your database so I can test this for myself. I only need the schema, and not the data. You can use a backup if it's small enough. Email to `simon@reversepoco.co.uk`

**Comment by VamsiKande** [comment]
Hi @sjh37 , I have sent you couple schemas for you test to simon@reversepoco.co.uk

I will try again replacing the ttinclude as suggested.

**Comment by VamsiKande** [comment]
I tried replacing the tt file, still no luck!

**Comment by sjh37** [comment]
Ok, group all your schema include filters into a single include filter, and do the same for all the other table/column include filters.
For example:

```c#
FilterSettings.SchemaFilters.Add(new RegexIncludeFilter("[Aa]dmin.*"));
FilterSettings.SchemaFilters.Add(new RegexIncludeFilter("[Rr]eport.*"));
FilterSettings.TableFilters.Add(new RegexIncludeFilter("^[Cc]ar.*"));
FilterSettings.TableFilters.Add(new RegexIncludeFilter("Rebel.*"));
```
Change to
```c#
FilterSettings.SchemaFilters.Add(new RegexIncludeFilter("[Aa]dmin.*|[Rr]eport.*"));
FilterSettings.TableFilters.Add(new RegexIncludeFilter("^[Cc]ar.*|Rebel.*"));
```

**Comment by VamsiKande** [comment]
Ok, can you also please send me an email of the tt file that you are testing with?

**Comment by VamsiKande** [comment]
I tested few other things, its not an issue related to the naming or using the Regex, its related to schema, with the schema that I sent in the email of [admin].[Convert_To_Demo]

commenting following line generates the stored proc, and uncommenting doesn't generate, Weird!!

select rtrim(@FirstName) + ' ' + @LastName as [Rep Name], @iClientID as [Client ID], @UserName as [Login]

also with the other stored proc [Report].[uspFreightInnovationsVolumeByDay], the select statement is having issue, commenting the select statement generates the stored proc, not sure what's causing that???

have you included the complete schema when you're testing??

**Comment by sjh37** [comment]
@VamsiKande Yes I included the contents of the stored procedures you sent me, however I don't have the tables they reference.

I did not create a separate .tt file, as I created a unit test for it.

Try moving the select
```c#
select rtrim(@FirstName) + ' ' + @LastName as [Rep Name], @iClientID as [Client ID], @UserName as [Login]
```
to the bottom of the stored procedure, so it's the last line before the `END`

Also at the top, you should also set
```c#
SET XACT_ABORT ON;
```
so if there is an error, it rolls back your transaction for you.

**Comment by VamsiKande** [comment]
@sjh37 tried moving the statement as suggested, that didn't work.

**Comment by sjh37** [comment]
Now released as part of v3.2.0
